Sources/Jetworking/Client/Client.swift

```diff
@@ -375,27 +385,36 @@ public final class Client {
         }
     }
 
+    private func handleSuccessVoid<ResponseType>(currentURLResponse: HTTPURLResponse, endpoint: Endpoint<ResponseType>?, data: Data?, completion: @escaping RequestCompletion<ResponseType>) {
+        if ResponseType.self is Void.Type {
+            return enqueue(completion(currentURLResponse, .success(() as! ResponseType)))
+        }
+
+        enqueue(completion(currentURLResponse, .failure(APIError.unexpectedError)))
+    }
+
+    func handleSuccessDecodable<ResponseType: Decodable>(currentURLResponse: HTTPURLResponse, endpoint: Endpoint<ResponseType>?, data: Data?, completion: @escaping RequestCompletion<ResponseType>) {
+        let decoder: Decoder = endpoint?.decoder ?? configuration.decoder
+
+        guard let decodedData = try? decodeResponse(ResponseType.self, with: decoder, from: data) else {
+            return enqueue(completion(currentURLResponse, .failure(APIError.decodingError)))
+        }
+        // Evaluate Header fields --> urlResponse.allHeaderFields
+
+        enqueue(completion(currentURLResponse, .success(decodedData)))
+    }
+
     private func enqueue(_ completion: @escaping @autoclosure () -> Void) {
         configuration.responseQueue.async {
             completion()
         }
     }
 
     /// Decodes an instance of the given type.
-    private func decodeResponse<ResponseType: Decodable>(_ dataType: ResponseType.Type, from data: Data?) throws -> ResponseType {
+    private func decodeResponse<ResponseType>(_ dataType: ResponseType.Type, with decoder: Decoder, from data: Data?) throws -> ResponseType where ResponseType: Decodable {
         guard let data = data else { throw APIError.decodingError }
 
-        return try configuration.decoder.decode(dataType, from: data)
-    }
-
-    /// Decodes an instance of the `Void` type. If a non-decodable type is given, throws an error instead.
-    ///
-    /// - IMPORTANT:
-    /// We let the method "overloaded" to allow `Void` type on return of a network response.
-    private func decodeResponse<ResponseType>(_ dataType: ResponseType.Type, from data: Data?) throws -> ResponseType {
-        guard ResponseType.self is Void.Type else { throw APIError.decodingError }
-
-        return () as! ResponseType
+        return try decoder.decode(dataType, from: data)
```

Comment:
Do we really want to `throw` here? I would rather have a dedicated error for this. What do you think? Maybe we should add a TODO here so that the one who is tackling error handling will not miss out on this.

Reply:
This functionality has moved to the ResponseHandler. Also its now decoded with a conditional `try`.

Reply:
Looks good 👍